FIAT/expansions.py

```diff
@@ -728,20 +736,24 @@ def compute_partition_of_unity(ref_el, pt, unique=True, tol=1E-12):
     :kwarg tol: the absolute tolerance.
     :returns: a list of (weighted) characteristic functions for each subcell.
     """
-    from sympy import Piecewise
+    from sympy import Piecewise, Or, Not
     sd = ref_el.get_spatial_dimension()
     top = ref_el.get_topology()
     # assert singleton point
     pt = pt.reshape((sd,))
 
-    # Compute characteristic function of each cell
+    # Compute the distance from the point to each subcell
+    dist = {cell: compute_l1_distance(ref_el, pt, entity=(sd, cell))
+            for cell in sorted(top[sd])}
+
+    # Compute characteristic function of each subcell
     otherwise = []
     masks = []
     for cell in sorted(top[sd]):
-        inside = compute_l1_distance(ref_el, pt, entity=(sd, cell)) < tol
-        masks.append(Piecewise(*otherwise, (1.0, inside), (0.0, True)))
+        nearest = Not(reduce(Or, (dist[cell] - dist[other] >= tol for other in dist if cell != other)))
```

Comment:
For N subcells, the number of comparisons is O(N^2), and this generates very slow sympy code. It would be nice the determine if we are the nearest cell with a single comparison.

Reply:
The l2 distance from the point to the nearest subcell must be equal to the distance from the point to the parent cell. I modified the computed l1 distance (wrt. barycentric coordinates) to match the Euclidean l2 distance for points in front of facets (simply scaling the only negative barycentric coordinate by the height of the simplex wrt. to the facet where this coordinate is zero, which simplifies to normalizing the columns of the Jacobian of the affine mapping from physical to barycentric coordinates).

The literal "corner cases" for points not directly in front of faces (more than one negative barycentric coordinate) are more tricky. The l1 distance to the parent cell will be larger than the l1 distance to the subcell if the subcell splits the parent vertices, because the subcell will use coordinate axes at a smaller angle.